sike.go

```diff
@@ -0,0 +1,174 @@
+	// SIKE keypair generation
+	// Outputs: SIKE secret key (message + secretkey + publickey)
+	//			SIKE public key (public key)
+
+	publicKey = new(SIDHPublicKeyBob)
+	SIKEsecretKey = new(SIKESecretKey)
+	var secretKey = new(SIDHSecretKeyBob)
+
+	//first randomly generate 32 byte s
+	_, err = io.ReadFull(rand, SIKEsecretKey.Scalar[:])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	//next, generate Encryptor secret-key and public-key
+	publicKey, secretKey, err = GenerateBobKeypair(rand)
+
+	//Copy secretKey and publicKey into SIKEsecretKey
+	SIKEsecretKey.SecretKey = secretKey
+	SIKEsecretKey.PublicKey = publicKey
+
+	return
+}
+
+func crypto_SIKE_encap(rand io.Reader, publicKey *SIDHPublicKeyBob) (cipherText *SIKECipherText, sharedSecret *SIKESharedSecret, err error) {
+	// SIKE encapsulation
+	// Input: publicKey
+	// Output: shared secret ss
+	//		   ciphertext message ct (publicKey + message)
+	cipherText = new(SIKECipherText)
+	sharedSecret = new(SIKESharedSecret)
+
+	const G = "0"
+	const H = "1"
+	const P = "2"
+	var ephemeral_sk = new(SIDHSecretKeyAlice)
+	var jinvariant [SharedSecretSize]byte
+	var h_ [MessageSize]byte
+	//var tmp [CiphertextSize + MessageSize]byte
+	var tmp = make([]byte, (CiphertextSize + MessageSize))
+
+	//Generate ephemeral secretKey G(m||pk) mod oA
+	_, err = io.ReadFull(rand, tmp[:MessageSize])
+	if err != nil {
+		return nil, nil, err
+	}
+
+	//Append publicKey to message and hash it
+	publicKey.ToBytes(tmp[MessageSize:])
+	sha3.CShakeSum256(ephemeral_sk.Scalar[:], tmp[:CiphertextSize], []byte(G))
+	//perform mod oA
+	ephemeral_sk.Scalar[47] = 0
+	ephemeral_sk.Scalar[46] &= 15 // clear high bits, so scalar < 2^372
+	ephemeral_sk.Scalar[0] &= 254 // clear low bit, so scalar is even
+
+	//Encryption
+	var tmp_pk = new(SIDHPublicKeyAlice)
+	//tmp_pk, ephemeral_sk, err = GenerateAliceKeypair(rand)
+
+	*tmp_pk = ephemeral_sk.PublicKey()
+	cipherText.PublicKey =  tmp_pk
+	jinvariant = ephemeral_sk.SharedSecret(publicKey)
+	sha3.CShakeSum256(h_[:], jinvariant[:], []byte(P))
+
+	for i := 0; i < MessageSize; i++ {
+		cipherText.Scalar[i] = tmp[i] ^ h_[i]
+	}
+
+	//Generate shared secret: ss = H(m||ct)
+	cipherText.PublicKey.ToBytes(tmp[MessageSize:])
+	copy(tmp[CiphertextSize:], cipherText.Scalar[:])
+	sha3.CShakeSum256(sharedSecret.Scalar[:], tmp[:], []byte(H))
+
+	return
+}
+
+func crypto_SIKE_decap(SIKEsecretKey *SIKESecretKey, cipherText *SIKECipherText) (sharedSecret *SIKESharedSecret){
+	// SIKE decapsulation
+	// Input: secretKey
+	//		  cipherText message
+	// Output: sharedSecret
+	sharedSecret = new(SIKESharedSecret)
+
+	const G = "0"
+	const H = "1"
+	const P = "2"
+	var ephemeral_sk = new(SIDHSecretKeyAlice)
+	var jinvariant [SharedSecretSize]byte
+	var h_ [MessageSize]byte
+	var c0_ = new(SIDHPublicKeyAlice)
+	var c0_bytes [PublicKeySize]byte
+	var c1_bytes [PublicKeySize]byte
+	//var tmp [CiphertextSize+MessageSize]byte
```

Comment:
commented out, also, please check spacing of comments and code blocks for readability